erts/emulator/beam/erl_term_hashing.c

```diff
@@ -0,0 +1,1724 @@
+#define FUNNY_NUMBER3  268435459
+#define FUNNY_NUMBER4  268436141
+#define FUNNY_NUMBER5  268438633
+#define FUNNY_NUMBER6  268437017
+#define FUNNY_NUMBER7  268438039
+#define FUNNY_NUMBER8  268437511
+#define FUNNY_NUMBER9  268439627
+#define FUNNY_NUMBER10 268440479
+#define FUNNY_NUMBER11 268440577
+#define FUNNY_NUMBER12 268440581
+#define FUNNY_NUMBER13 268440593
+#define FUNNY_NUMBER14 268440611
+
+static Uint32
+hash_binary_bytes(Eterm bin, Uint sz, Uint32 hash)
+{
+    byte* ptr;
+    Uint bitoffs;
+    Uint bitsize;
+
+    ERTS_GET_BINARY_BYTES(bin, ptr, bitoffs, bitsize);
+    if (bitoffs == 0) {
+        while (sz--) {
+            hash = hash*FUNNY_NUMBER1 + *ptr++;
+        }
+        if (bitsize > 0) {
+            byte b = *ptr;
+
+            b >>= 8 - bitsize;
+            hash = (hash*FUNNY_NUMBER1 + b) * FUNNY_NUMBER12 + bitsize;
+        }
+    } else {
+        Uint previous = *ptr++;
+        Uint b;
+        Uint lshift = bitoffs;
+        Uint rshift = 8 - lshift;
+
+        while (sz--) {
+            b = (previous << lshift) & 0xFF;
+            previous = *ptr++;
+            b |= previous >> rshift;
+            hash = hash*FUNNY_NUMBER1 + b;
+        }
+        if (bitsize > 0) {
+            b = (previous << lshift) & 0xFF;
+            previous = *ptr++;
+            b |= previous >> rshift;
+
+            b >>= 8 - bitsize;
+            hash = (hash*FUNNY_NUMBER1 + b) * FUNNY_NUMBER12 + bitsize;
+        }
+    }
+    return hash;
+}
+
+Uint32 make_hash(Eterm term_arg)
+{
+    DECLARE_WSTACK(stack);
+    Eterm term = term_arg;
+    Eterm hash = 0;
+    unsigned op;
+
+#define MAKE_HASH_TUPLE_OP      (FIRST_VACANT_TAG_DEF)
+#define MAKE_HASH_TERM_ARRAY_OP (FIRST_VACANT_TAG_DEF+1)
+#define MAKE_HASH_CDR_PRE_OP    (FIRST_VACANT_TAG_DEF+2)
+#define MAKE_HASH_CDR_POST_OP   (FIRST_VACANT_TAG_DEF+3)
+
+    /* 
+    ** Convenience macro for calculating a bytewise hash on an unsigned 32 bit 
+    ** integer.
+    ** If the endianess is known, we could be smarter here, 
+    ** but that gives no significant speedup (on a sparc at least) 
+    */
+#define UINT32_HASH_STEP(Expr, Prime1)                                        \
+        do {                                                                \
+            Uint32 x = (Uint32) (Expr);                                        \
+            hash =                                                        \
+                (((((hash)*(Prime1) + (x & 0xFF)) * (Prime1) +                \
+                ((x >> 8) & 0xFF)) * (Prime1) +                                \
+                ((x >> 16) & 0xFF)) * (Prime1) +                        \
+                 (x >> 24));                                                \
+        } while(0)
+
+#define UINT32_HASH_RET(Expr, Prime1, Prime2)           \
+        UINT32_HASH_STEP(Expr, Prime1);                        \
+        hash = hash * (Prime2);                                \
+        break
```

Comment:
While at I, I suggest that you clean up the arrangement of the backslashes.

Reply:
I am not sure what you mean with this:
`My first attempt did that, with a check of all extensions at startup so that it would fail immediately with an error message telling the user to rebuild with a certain flag, but after discussing it with @rickard-green we decided to leave it entirely to the user.`

At startup of the Erlang VM we test if the instructions we intend to use (chosen at configure and defaults) actually is supported on the CPU where we are running. We then terminate in an understandable way.

With "leave entirely to the user" I don't understand what you mean. Should it just crash later at some random time depending on what the Erlang system is doing? Should it crash with the user reporting a suspected bug to us?

I would also think that we should clearly state that we don't guarantee that our binaries built on or for one system works on other systems without the user taking care of that by selecting the appropriate options at configure time. I also assume that cross builds has a role to play here. A cross build it intended for a specific OS and architecture, it is up to the user to get this right.

Reply:
I just talked to @rickard-green and know more now. You don't need to answer my previous comment